driver-core/src/main/java/com/datastax/driver/core/HostConnectionPool.java

```diff
@@ -271,12 +275,7 @@ public void returnConnection(PooledConnection connection) {
             return;
         }
 
-        int inFlight = connection.inFlight.decrementAndGet();
-
-        if (trash.contains(connection)) {
-            if (inFlight == 0 && trash.remove(connection))
-                close(connection);
-        } else {
+        if (!trash.contains(connection)) {
```

Comment:
Yes. This comes from the previous implementation, and it didn't occur to me to change it. It doesn't make a big difference, but it's more logical since {{state}} is what we use to do the CAS.